Enterprise-CMCS/macpro-mdct-hcbs
# Add authorization to the API

### Description
This PR adds permission checks to the API, but it also restructures it.



Currently in main, the API flow looks like this:
![handler_before](https://github.com/user-attachments/assets/69223455-25c7-4c7c-bf31-4307aef45d07)
1. handler-lib verifies the signature on the JWT, confirming that this request is from a valid user
2. handler-lib calls the appropriate lambda, passing in the entire Event
3. The lambda reads the path parameters, such as state and reportType, from the Event
4. The lambda reads the user data from the JWT
   * This includes a one-size-fits-all permission check; that either the user is an admin or they are from the correct state
5. The lambda does whatever actual operation it's supposed to

This PR changes the flow, to this:
![handler_after](https://github.com/user-attachments/assets/26511870-295e-4f60-aa11-0095875ce62b)
1. handler-lib verifies the signature on the JWT _while also_ reading the user data
2. handler-lib reads the path parameters from the Event
   * Making use of whichever Parameter Parser it receives from the endpoint
3. handler-lib calls the appropriate lambda, passing in the refined Request
   * (I apologize if the layout of the diagram makes the sequence unclear)
5. The lambda performs whichever permission checks are appropriate
6. The lambda does whatever actual operation it's supposed to

I will admit this feels a little awkward, with the parser being passed up from the endpoint to handler-lib. I'll also note that this is among the first uses of a generic type parameter in our codebase: The return type of the parser (such as `parseReportParameters`) flows through to determine the type details of the Request (such as the presence/absence of `request.parameters.id`).

***Question for Reviewers:*** Is this tradeoff worth it? What we gain is the removal of a couple lines of parameter-parsing boilerplate, at the top of every endpoint lambda. What we pay is this increase in complexity. Is the API flow still easy enough to follow? Will the generic type parameters be a hurdle for future maintainers?

### Related ticket(s)
CMDCT-4005

---
### How to test
1. Log in with `stateuser@test.com` at https://d1x4yjcaf95sje.cloudfront.net/
2. Create a report - note that it succeeds
3. Make some changes to the report data - note that it saves successfully
4. Go back to the QM dashboard, and open the browser console
5. Change the state in the URL, and navigate there
7. Note the error in the console. This came from our new authorization check.


### Important updates
This PR moves parameter parsing to handler-lib! This is a major pattern change, for the whole API.

---
### Author checklist
